hal/src/nRF52840/spi_hal.cpp

```diff
@@ -299,6 +299,9 @@ void HAL_SPI_Begin_Ext(HAL_SPI_Interface spi, SPI_Mode mode, uint16_t pin, void*
             // m_spi_map[spi].ss_pin = PIN_INVALID;
         }
     } else {
+        if (pin >= TOTAL_PINS) {
```

Comment:
Why are we removing all the checks for `TOTAL_PINS`? Especially in HAL, which will break backwards compatibility.

It's entirely possible to pass `PIN_INVALID` as the pin or some other invalid number.

Reply:
I'm don't how it breaks the backwards compatibility. By moving the check routine from wiring to HAL just gaining the time that the wiring API acquiring/releasing the lock. Even if user pass invalid pin number, the HAL can still handle it correctly. So from user's perspective, the result calling the wiring API makes no difference. The root cause of the issue is that the `begin()` APIs are not type safe, so when user calling `SPI.begin(SPI_MODE_MASTER)`, it will call the `SPIClass::begin(uint16_t ss_pin)` instead. To fix it I just assign a default value `SPI_DEFAULT_SS` for the second argument of the `SPIClass::begin(SPI_Mode mode, uint16_t ss_pin)` API, so that it can invoke the corresponding overloaded function. If we don't move the check routine to HAL, the wiring API will return immediately if calling `SPI.begin(SPI_MODE_MASTER)`.

Reply:
You are right. I mistakenly thought we also removed this check here, not added it :smile:

:+1: